rfcs/20190821-nodejs-saved-model.md

```diff
@@ -0,0 +1,401 @@
+*   Enable the ability to inspect the SavedModel signatureDefs in Node.js with protocol buffers in JavaScript (stage 2 feature).
+
+### Non-goals
+
+*   Enable support for training a SavedModel
+*   Enable support for exporting a SavedModel
+
+## **Motivation**
+
+[TensorFlow.js](https://www.tensorflow.org/js/) brings TensorFlow into the JavaScript world. Currently users could use [tfjs-converter](https://github.com/tensorflow/tfjs-converter) to convert TensorFlow SavedModel and TensorFlow Hub module to JS friendly format and to run inference through TensorFlow.js. The tfjs-converter tool requires developers to install Python TensorFlow and some parameters/configurations, which we have noticed users are struggling with.
+
+The [tfjs-node](https://github.com/tensorflow/tfjs/tree/master/tfjs-node) project provides native TensorFlow execution in Node.js environment through TensorFlow C library under the hood. It provides the same API (140+ ops) as [TensorFlow.js](https://js.tensorflow.org/api/latest/), which is a subset of the TensorFlow ops (900+). The TensorFlow C library actually supports more ops and APIs beyond ops.
+
+Here there is an opportunity to support native SavedModel execution in Node.js so that 1) tfjs-node can support models which contain ops that are not supported in tfjs yet, and 2) users do not need to go through extra conversion of the model.
+
+
+## **Design Proposal**
+
+
+### User-facing code
+
+This project will provide a new API `tf.node.loadSavedModel` to load a SavedModel as a new class `TFSavedModel` in Node.js which can be used by the user to execute the SavedModel for inference.
+
+The loadSavedModel API takes a `path` param, which is the absolute path to the SavedModel directory. It returns a `TFSavedModel` object, implementing [InferenceModel](https://github.com/tensorflow/tfjs/blob/81225adc2fcf6fcf633b4119e4b89a3bf55be824/tfjs-core/src/model_types.ts#L36) class.
+
+
+```
+const savedMode = tf.node.loadSavedModel(__dirname + 'saved_model_dir');
+```
+
+
+The returned TFSavedModel object has a `predict()` function to execute the SavedModel serving_default signature for inference. The param of this predict() function would be a single tensor if there is single input for the model, an array of tensors or named tensor map if the model has multiple inputs.
+
+The TFSaveModel object also has an `execute()` function to execute the inference for the input tensors and return activation values for specified output node names.
+
+
+```
+const input = tensor1d([123], 'int32');
+// Execute default signatureDef of the SavedModel
+const output = savedMode.predict([input_tensors]);
+
+// Execute multiple signatureDefs of the SavedModel
+const output = savedMode.execute({'input_op_names':input_tensors}, ['output_op_names']);
+```
+
+
+The TFSavedModel object also has a `delete()` function to free the SavedModel related memory.
```

Comment:
Is this call considered a good practice or a requirement?

Reply:
It's not a hard requirement, but it will help with memory management, especially for servers that keeps running.